libp2p/protocols/mix/entry_connection.nim

```diff
@@ -26,32 +39,126 @@ chronicles.formatIt(MixEntryConnection):
 method readOnce*(
     s: MixEntryConnection, pbytes: pointer, nbytes: int
 ): Future[int] {.async: (raises: [CancelledError, LPStreamError]), public.} =
-  # TODO: implement
-  raise newLPStreamEOFError()
+  if s.isEof:
+    raise newLPStreamEOFError()
+
+  try:
+    await s.replyReceivedFut
+    s.isEof = true
+    if s.cached.len == 0:
+      raise newLPStreamEOFError()
+  except CancelledError as exc:
+    raise exc
+  except LPStreamEOFError as exc:
+    raise exc
+  except CatchableError as exc:
+    raise (ref LPStreamError)(msg: "error in readOnce: " & exc.msg, parent: exc)
+
+  let toRead = min(nbytes, s.cached.len)
+  copyMem(pbytes, addr s.cached[0], toRead)
+  s.cached = s.cached[toRead ..^ 1]
+  return toRead
 
 method readExactly*(
     s: MixEntryConnection, pbytes: pointer, nbytes: int
 ): Future[void] {.async: (raises: [CancelledError, LPStreamError]), public.} =
-  # TODO: implement
-  raise newLPStreamEOFError()
+  ## Waits for `nbytes` to be available, then read
+  ## them and return them
+  if s.atEof:
+    var ch: char
+    discard await s.readOnce(addr ch, 1)
+    raise newLPStreamEOFError()
+
+  if nbytes == 0:
+    return
+
+  logScope:
+    s
+    nbytes = nbytes
+    objName = s.objName
+
+  var pbuffer = cast[ptr UncheckedArray[byte]](pbytes)
+  var read = 0
+  while read < nbytes and not (s.atEof()):
+    read += await s.readOnce(addr pbuffer[read], nbytes - read)
+
+  if read == 0:
+    doAssert s.atEof()
+    trace "couldn't read all bytes, stream EOF", s, nbytes, read
+    # Re-readOnce to raise a more specific error than EOF
+    # Raise EOF if it doesn't raise anything(shouldn't happen)
+    discard await s.readOnce(addr pbuffer[read], nbytes - read)
+
+    raise newLPStreamEOFError()
+
+  if read < nbytes:
+    trace "couldn't read all bytes, incomplete data", s, nbytes, read
+    raise newLPStreamIncompleteError()
```

Comment:
could you explain this? every `MixEntryConnection.readLp(size)` must read `size` bytes? 
we usually had this `size` as maximum bytes, but not necessarily exactly.